crates/trie/sparse-parallel/src/trie.rs

```diff
@@ -885,6 +974,198 @@ impl SparseSubtrie {
         current_level == child_level
     }
 
+    /// Updates or inserts a leaf node at the specified key path with the provided RLP-encoded
+    /// value.
+    ///
+    /// If the leaf did not previously exist, this method adjusts the trie structure by inserting
+    /// new leaf nodes, splitting branch nodes, or collapsing extension nodes as needed.
+    ///
+    /// # Returns
+    ///
+    /// Returns the `Ok` if the update is successful.
+    /// If a split branch was added this is returned as well, along with its path.
+    ///
+    /// Note: If an update requires revealing a blinded node, an error is returned if the blinded
+    /// provider returns an error.
+    pub fn update_leaf(
+        &mut self,
+        full_path: Nibbles,
+        value: Vec<u8>,
+        provider: impl BlindedProvider,
+    ) -> SparseTrieResult<()> {
+        debug_assert!(full_path.starts_with(&self.path));
+        let existing = self.inner.values.insert(full_path, value);
+        if existing.is_some() {
+            // trie structure unchanged, return immediately
+            return Ok(())
+        }
+
+        // Here we are starting at the root of the subtrie, and traversing from there.
+        let mut current = Some(self.path);
+        while let Some(current_path) = current {
+            match self.next_node(current_path, &full_path, &provider)? {
+                LeafUpdateStep::Continue { next_node } => {
+                    current = Some(next_node);
+                }
+                LeafUpdateStep::Complete { .. } | LeafUpdateStep::NoOp => {
+                    current = None;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Processes the current node, returning what to do next in the leaf update process.
+    ///
+    /// This will add or update any nodes in the trie as necessary.
+    ///
+    /// Returns a `LeafUpdateStep` containing the next node to process (if any) and
+    /// the paths of nodes that were inserted during this step.
+    fn next_node(
```

Comment:
just to be clear that it belongs to the `update` part of the sparse trie, and also that it modifies the structure, not just iterates over it
```suggestion
    fn update_next_node(
```

Reply:
yep makes sense